### DISCUSSION TITLE
How to Make a Script/Task to Set RTC Regularly?

### DISCUSSION THREAD

**Discussion by AlexLandherr** [PiSupply/PiJuice#892] [Q&A] [answered]
I have a NEO-M8T GNSS Timing HAT together with a PiJuice. I use the GNSS HAT to set the system time with PPS (Pulse Per Second).

This results in the Pi having accurate time to within ~1 microsecond when idling.

My question as in the title is how I can set up a task/script to regularly (every 24 hours say) read the very accurate system time and write it to the RTC of the PiJuice?

**Comment by tvoverbeek** [comment, marked as answer]
I am assuming you set the Linux system time from the GNSS HAT.
Linux internal system time is in UTC.
Just run e.g. a 24h cron job which copies the system time to the RTC: `sudo hwclock --systohc `
You might have to also use the --rtc option if your GNSS HAT also has a /dev/rtcx
Of course the device driver for the PiJuice RTC has to be loaded.
See `man hwclock` for more details.
Do not expect microsecond precision from the PiJuice RTC.

**Reply by AlexLandherr** [reply]
I don't expect such precision from the RTC, it just needs to play the part of a holdover clock.

And my particular GNSS HAT doesn't have an RTC.

**Reply by AlexLandherr** [reply]
So the cron job would look something like this?:
`@daily sudo hwclock --systohc`

**Reply by tvoverbeek** [reply]
Yes, provided the RTC driver for the PiJuice is loaded.

**Reply by AlexLandherr** [reply]
> Yes, provided the RTC driver for the PiJuice is loaded.
> 
> 

Which guide shows how to do that? I've looked but probably missed it.

**Reply by tvoverbeek** [reply]
If you have the PiJuice HAT there is probably a conflict with the GNSS HAT which also shoud have an EEPROM at address 0x50.
In the PiJuice CLI or GUI change the EEPROM address to 0x52. 
You have to add the line `dtoverlay=i2c-rtc,ds1307` to /boot/config.txt.
After boot you can check if /dev/rtc0 exists.

**Reply by AlexLandherr** [reply]
> If you have the PiJuice HAT there is probably a conflict with the GNSS HAT which also shoud have an EEPROM at address 0x50.
> 
> In the PiJuice CLI or GUI change the EEPROM address to 0x52. 
> 
> You have to add the line `dtoverlay=i2c-rtc,ds1307` to /boot/config.txt.
> 
> After boot you can check if /dev/rtc0 exists. 

Thanks! Will try it when I get the time.

**Reply by ArjenR49** [reply]
I don't have any possibly conflicting HAT, and no dtoverlay=i2c-rtc,ds1307 in config.txt. 
Yet the RTC device is present:

```
pi@RPI-HUB:~ $ ls -la /dev/rtc*
lrwxrwxrwx 1 root root      4 20 mrt 15:34 /dev/rtc -> rtc0
crw------- 1 root root 252, 0 20 mrt 15:34 /dev/rtc0
pi@RPI-HUB:~ $ 

```
It was my understanding that the RTC overlay is added through the EEPROM. I've seen a text to that effect, which was intended to be compiled into a binary file and loaded into the EEPROM. I do not know the current content of my PiJuice EEPROM (maybe can be read and decompiled?), but the RTC works.

**Comment by tvoverbeek** [comment]
Yes, the RTC overlay is in the EEPROM.
However it loads the RTC module for the ds1339. This is identical to the ds1307, except it includes the alarm functions which can
then also be controlled by the OS. When connected to the internet the regular system time updates by systemd-timesyncd also clear any set alarms set via the PiJuice CLI or GUI. Therefore the recommendation to use `dtoverlay=i2c-rtc,ds1307` in /boot/config.txt to avoid the OS messing around with the alarm settings.

To dump the contents of the eeprom, see https://github.com/raspberrypi/hats/tree/master/eepromutils.
Use eepflash to read the eeprom and eepdump to make it human readable.
